mne/time_frequency/tfr.py

```diff
@@ -245,6 +249,239 @@ def _cwt(X, Ws, mode="same", decim=1, use_fft=True):
+    # Check frequencies
+    if not isinstance(frequencies, (list, np.ndarray)):
+        raise ValueError('frequencies must be an array-like.')
+    frequencies = np.asarray(frequencies, dtype=float)
+    if frequencies.ndim != 1:
+        raise ValueError('frequencies must be of shape n_freqs.')
+
+    # Check sfreq
+    if not isinstance(sfreq, (float, int)):
+        raise ValueError('sfreq must be a float or an int.')
+    sfreq = float(sfreq)
+
+    # Check output
+    allowed_ouput = ('complex', 'power', 'phase',
+                     'avg_power_phaselock', 'avg_power', 'avg_phaselock')
+    if output not in allowed_ouput:
+        raise ValueError("Unknown output type.")
+
+    if method not in ('mtm', 'morlet'):
+        raise ValueError('method must be "morlet" or "mtm"')
+
+    # Default zero_mean = True if mtm else False
+    zero_mean = method == 'mtm' if zero_mean is None else zero_mean
+    if not isinstance(zero_mean, bool):
+        raise ValueError('')
+    frequencies = np.asarray(frequencies)
+
+    # XXX Can we compute single-trial phases with mtm?
+    if (method == 'mtm') and (output == 'phase'):
+        raise NotImplementedError(
+            'This function is not optimized to compute the phase using the '
+            'mtm method. Use np.angle of the complex output instead.')
+
+    # Setup wavelet
+    if method == 'morlet':
+        W = morlet(sfreq, frequencies, n_cycles=n_cycles, zero_mean=zero_mean)
+        Ws = [W]  # to have same dimensionality as the 'mtm' case
+        if time_bandwidth != 4.0:
+            raise ValueError('time_bandwidth only applies to "mtm" method.')
+    elif method == 'mtm':
+        Ws = dpss_wavelet(sfreq, frequencies, n_cycles=n_cycles,
+                          time_bandwidth=time_bandwidth, zero_mean=zero_mean)
+
+    # Check wavelets
+    if len(Ws[0][0]) > epoch_data.shape[2]:
+        raise ValueError('At least one of the wavelets is longer than the '
+                         'signal. Use a longer signal or shorter wavelets.')
+
+    # Initialize output
+    decim = _check_decim(decim)
+    n_freqs = len(frequencies)
+    n_epochs, n_chans, n_times = epoch_data[:, :, decim].shape
+    if output in ('power', 'phase', 'avg_power', 'avg_phaselock'):
+        dtype = np.float
+    elif output in ('complex', 'avg_power_phaselock'):
+        # avg_power_phaselock is stored as power + 1i * phase_lock to keep a
+        # simple dimensionality
+        dtype = np.complex
+
+    if 'avg_' in output:
+        out = np.empty((n_chans, n_freqs, n_times), dtype)
+    else:
+        out = np.empty((n_chans, n_epochs, n_freqs, n_times), dtype)
+
+    # Parallel computation
+    parallel, my_cwt, _ = parallel_func(_time_frequency_loop, n_jobs)
+
+    # Parallization is applied across channels.
+    tfrs = parallel(
+        my_cwt(channel, Ws, output, use_fft, 'same', decim)
+        for channel in epoch_data.transpose(1, 0, 2))
+
+    # FIXME: to avoid overheads we should use np.array_split()
```

Comment:
> I agree. If someone hits optimization issues maybe he/she can benchmark the two routes options.

But it used to iterate the other way, right? In other words, your PR changes the way the parallelization is done, and therefore could introduce speed regressions for some use cases, and that hasn't been checked...?

>  In the worse case scenario, we pickle 306 times don't we?

Ahh I thought you were iterating over `n_epochs * n_channels` 1D (time) arrays, but I see now you're iterating over `n_channels` 2D (epoch x time) arrays. So yeah 306. In that case, I suppose the iteration order shouldn't matter too much, assuming there isn't a huge difference in the number of channels vs the number of epochs people will tend to have. Is that what you have in mind?